### PR TITLE
[chat] - send message

### PR DESCRIPTION
**Pull request by florianlenz**
added logic to send chat message

### CODE REVIEW

File: chat/chat.go [hunk 1]
```diff
@@ -1,105 +1,36 @@
...
-
-	if err := preKeyB.PublicPart.Sign(*c.km); err != nil {
-		return PanthalassaPreKeyBundle{}, err
-	}
-
-	return preKeyB, nil
-
-}
+const (
+	SignedPreKeyValidTimeFrame = time.Second * 3600 * 24 * 60
```
Comment (borjantrajanoski) [inline chat/chat.go]: SignedPreKeyValidTimeFrame will likely be a relatively long period of time which can be simplified by using 
`SignedPreKeyValidTimeFrame = time.Hour * 24 * 60`

File: chat/prekey/key_test.go [hunk 2]
```diff
@@ -72,3 +73,10 @@ func TestPreKey_ToProtobufAndBack(t *testing.T) {
 	require.Equal(t, hex.EncodeToString(pp.Key), hex.EncodeToString(k.PublicKey[:]))
 
 }
+
+func TestPreKey_OlderThen(t *testing.T) {
```
Comment (borjantrajanoski) [inline chat/prekey/key_test.go]: OlderThen should be renamed to OlderThan to be grammatically correct.

File: chat/prekey/key_test.go [hunk 3]
```diff
@@ -72,3 +73,10 @@ func TestPreKey_ToProtobufAndBack(t *testing.T) {
...
+	k := PreKey{
+		time: time.Now().Truncate(time.Second * 10),
+	}
+	require.False(t, k.OlderThen(time.Second*5))
```
Comment (borjantrajanoski) [inline chat/prekey/key_test.go]: OlderThen should be renamed to OlderThan to be grammatically correct.

File: chat/send_message.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+	}
+
+	// fetch signed pre key from storage
+	signedPreKey, err := fetchSignedPreKey(receiver)
+	if err != nil {
+		return handleSendError(err)
+	}
+
+	// check if signed pre key expired
+	expired := signedPreKey.OlderThen(SignedPreKeyValidTimeFrame)
```
Comment (borjantrajanoski) [inline chat/send_message.go]: OlderThen should be renamed to OlderThan to be grammatically correct.

File: chat/utils.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	validSig, err := signedPreKey.VerifySignature(idPubKey)
+	if err != nil {
+		return err
+	}
+	if !validSig {
+		return errors.New("signed pre key signature is invalid")
+	}
+
+	// check if signed pre key didn't expire
+	expired := signedPreKey.OlderThen(SignedPreKeyValidTimeFrame)
```
Comment (borjantrajanoski) [inline chat/utils.go]: OlderThen should be renamed to OlderThan to be grammatically correct.

File: chat/prekey/key.go [hunk 6]
```diff
@@ -119,3 +119,8 @@ func FromProtoBuf(preKey pb.PreKey) (PreKey, error) {
 	copy(p.identityPublicKey[:], preKey.IdentityKey[:32])
 	return p, nil
 }
+
+// check if pre key is older than given date
+func (p PreKey) OlderThen(past time.Duration) bool {
```
Comment (borjantrajanoski) [inline chat/prekey/key.go]: OlderThen should be renamed to OlderThan to be grammatically correct.

File: chat/chat.go (hunk 1 above)
Comment (florianlenz) [reply chat/chat.go]: Done

File: chat/prekey/key_test.go (hunk 2 above)
Comment (florianlenz) [reply chat/prekey/key_test.go]: Done

File: chat/prekey/key_test.go (hunk 3 above)
Comment (florianlenz) [reply chat/prekey/key_test.go]: Done

File: chat/utils.go (hunk 5 above)
Comment (florianlenz) [reply chat/utils.go]: done

File: chat/prekey/key.go (hunk 6 above)
Comment (florianlenz) [reply chat/prekey/key.go]: done

**Review by borjantrajanoski — APPROVED** [review]
Looks good 👍

**Pull request merged**